src/web/build_details.rs

```diff
@@ -107,4 +121,67 @@ mod tests {
             Ok(())
         });
     }
+
+    #[test]
+    fn s3_build_logs() {
+        wrapper(|env| {
+            env.fake_release()
+                .name("foo")
+                .version("0.1.0")
+                .builds(vec![FakeBuild::default().s3_build_log("A build log")])
+                .create()?;
+
+            let page = kuchiki::parse_html().one(
+                env.frontend()
+                    .get("/crate/foo/0.1.0/builds")
+                    .send()?
+                    .text()?,
+            );
+
+            let node = page.select("ul > li a.release").unwrap().next().unwrap();
+            let attrs = node.attributes.borrow();
+            let url = attrs.get("href").unwrap();
+
+            let page = kuchiki::parse_html().one(env.frontend().get(url).send()?.text()?);
+
+            let log = page.select("pre").unwrap().next().unwrap().text_contents();
+
+            assert!(log.contains("A build log"));
+
+            Ok(())
+        });
+    }
+
+    #[test]
+    fn both_build_logs() {
+        wrapper(|env| {
+            env.fake_release()
+                .name("foo")
+                .version("0.1.0")
+                .builds(vec![FakeBuild::default()
+                    .s3_build_log("A build log")
+                    .db_build_log("Another build log")])
+                .create()?;
+
+            let page = kuchiki::parse_html().one(
+                env.frontend()
+                    .get("/crate/foo/0.1.0/builds")
+                    .send()?
+                    .text()?,
+            );
+
+            let node = page.select("ul > li a.release").unwrap().next().unwrap();
+            let attrs = node.attributes.borrow();
+            let url = attrs.get("href").unwrap();
+
+            let page = kuchiki::parse_html().one(env.frontend().get(url).send()?.text()?);
+
+            let log = page.select("pre").unwrap().next().unwrap().text_contents();
+
+            // Relatively arbitrarily the DB is prioritised
+            assert!(log.contains("Another build log"));
```

Comment:
That seems fine, yeah - the alternative is to check each time we get something from the database whether there's also something in S3, which is expensive and shouldn't be necessary. Maybe we could add a mode that double checks it's deduplicated but that also sounds expensive.